src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Headers.cs

```diff
@@ -287,4 +289,152 @@ public async Task SendAsync_WithZeroLengthHeaderName_Throws()
+
+        private static bool AllowLatin1Headers(string useAppCtxSwitchInner, string switchValueInner)
+        {
+            bool switchValue = bool.Parse(switchValueInner);
+            if (bool.Parse(useAppCtxSwitchInner))
+            {
+                AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.AllowLatin1Headers", switchValue);
+            }
+            else
+            {
+                Environment.SetEnvironmentVariable("DOTNET_SYSTEM_NET_HTTP_SOCKETSHTTPHANDLER_ALLOWLATIN1HEADERS", switchValueInner);
+            }
+            return switchValue;
+        }
+
+        // bool useAppCtxSwitch, bool switchValue, bool utf8
+        public static readonly TheoryData<bool, bool, bool> NonAsciiTestConfigurations = new TheoryData<bool, bool, bool>()
+        {
+            { true, false, false },
+            { true, false, true },
+            { true, true, false },
+            { true, true, true },
+            { false, true, false },
+        };
+
+        // Since RemoteExecutor cannot invoke delegates defined in abstract types, we should define the actual test cases in derived classes.
+        protected int SendAsync_SendNonAsciiCharacters_Inner(string useAppCtxSwitchInner, string switchValueInner, string unicodeInner)
+        {
+            bool allowLatin1 = AllowLatin1Headers(useAppCtxSwitchInner, switchValueInner);
+            bool unicode = bool.Parse(unicodeInner);
+            bool expectSuccess = allowLatin1 && !unicode;
+
+            (string name, string[] values)[] headers = GetNonAsciiTestHeaders(unicode);
+
+            LoopbackServerFactory.CreateClientAndServerAsync(
+                async uri =>
+                {
+                    using var requestMessage = new HttpRequestMessage(HttpMethod.Get, uri) { Version = VersionFromUseHttp2 };
+
+                    foreach ((string name, string[] values) in headers)
+                    {
+                        requestMessage.Headers.Add(name, values);
+                    }
+
+                    using HttpClient client = CreateHttpClient();
+
+                    if (expectSuccess)
+                    {
+                        await client.SendAsync(requestMessage);
+                    }
+                    else
+                    {
+                        await Assert.ThrowsAsync<HttpRequestException>(() => client.SendAsync(requestMessage));
+                    }
+                },
+                async server =>
+                {
+                    HttpRequestData requestData = null;
+                    try
+                    {
+                        requestData = await server.HandleRequestAsync();
+                    }
+                    catch (Exception)
+                    {
+                        if (expectSuccess)
+                            throw;
+                    }
+
+                    if (!expectSuccess)
```

Comment:
It doesn't throw in this case?

Reply:
We swallow the exception in this case.

I can also imagine a case when `expectSuccess == false` but `await server.HandleRequestAsync()` does not throw, so returning here seemed the most robust solution here.